Fixed point multiplication improvements for AArch64

giuseros opened the pull request:
### RFC
This PR is based on the following RFC: https://discuss.tvm.ai/t/rfc-using-arm-intrinsics-to-implement-fixed-point-multiplication-in-tvm

### High level description of the submission
The idea is to create a TIR intrinsic `fixed_point_multiply` that can be overloaded (by the means of `tvm.target.intrin.register_intrin_rule`) and use AArch64 intrinsics (other vendors can provide their own implementation)

The TIR default intrinsic is registered in: `tvm/src/target/intrin_rule.cc`
The overload for AArch64 lives in `tvm/topi/python/topi/arm_cpu/tensor_intrin.py`

giuseros commented:
cc @tqchen @anijain2305 @kparzysz-quic

giuseros commented:
Following the discussion in the [RFC](https://discuss.tvm.ai/t/rfc-using-arm-intrinsics-to-implement-fixed-point-multiplication-in-tvm) I redesigned the submission by introducing a general `qmuls` intrinsic:
* Relay uses `qmuls` to implement `fixed_point_multiply`
* The new intrinsic is more general, as it accepts the Q-ness of the input values to be multiplied

The structure of the submission remains the same.

giuseros commented:
@anijain2305 @kparzysz-quic @tqchen Any update on this?

kparzysz-quic commented on include/tvm/tir/op.h:
```diff
@@ -552,6 +552,24 @@ TVM_DLL PrimExpr trunc(PrimExpr x);
+ * (i.e., round(x.1) = x and round (x.5) = x+1)
+ * \param x first Q-number
+ * \param y second Q-number
+ * \param q Q-ness of x and y
```
You could say that `q` is the number of fractional bits in `x` and `y`.  We should also require that `q > 0`, or otherwise the code in [intrin_rule.cc at lines 151-152](https://github.com/apache/incubator-tvm/pull/5980/files#diff-df842e89e6e7af56d4566cee5d38f7f3R151-R152) won't work.

anijain2305 replied:
Agreed, number of fractional bits is better description

anijain2305 replied:
Are number of fractional bits same for x and y and thats why we need only input? Lets make the description more clear.
IIUC, one can think of it using this op to perform something like Q1.31  * Q2.30. But, I think this op is restrictive than that. If it is, then lets mention it.

giuseros replied:
I added the constraint of `q>0` in the doc string of the function (both in python and c++). The problem is that this intrinsic is vectorized and `q` becomes a vector. In order to check `q>0` within the intrinsic, I would need to go through a reduction (which seems a bit expensive for a simple check).

kparzysz-quic commented on python/tvm/tir/op.py:
```diff
@@ -965,6 +965,34 @@ def popcount(x):
+    y : PrimExpr
+        Second Q-number
+    q : PrimExpr
+        Q-ness of x and y
```
Same here.

kparzysz-quic commented on src/target/intrin_rule.cc:
```diff
@@ -115,6 +115,51 @@ TVM_REGISTER_GLOBAL("tvm.intrin.rule.default.isinf")
+
+      // 1) Calculating the integer multiplier and integer shift
+      PrimExpr zero = make_const(s.dtype(), 0);
+      PrimExpr left_shift = tir::Select((s > zero), s, zero);
+      PrimExpr right_shift = tir::Select(s > zero, zero, -s);
```
You could remove the parentheses in `(s > zero)`, so that it looks consistent in the two lines.

anijain2305 commented on include/tvm/relay/attrs/transform.h:
```diff
@@ -298,6 +298,17 @@ struct ClipAttrs : public tvm::AttrsNode<ClipAttrs> {
+  int32_t shift;
+
+  TVM_DECLARE_ATTRS(FixedPointMultiplyAttrs, "relay.attrs.FixedPointMultiplyAttrs") {
+    TVM_ATTR_FIELD(multiplier).describe("Integer multiplier.");
```
Nit, but lets remove the period at the end to be consistent with others.
might be good to describe the multiplier and shift briefly

anijain2305 commented on include/tvm/tir/builtin.h:
```diff
@@ -92,6 +92,14 @@ TVM_DLL const Op& shift_right();
+ * The default rounding rule is to the nearest value, rounding half up
+ * (i.e., round(x.1) = x and round (x.5) = x+1)
+ */
+TVM_DLL const Op& qmuls();
```
We should come up with a better name. Currently, `qmuls` seems vague.
Not sure what `q` and `s` stand for a person not familiar with Q numbers.

Why not use the same `fixed_point_multiply`?

giuseros replied:
Should we use this thread to discuss naming? The problem with using `fixed_point_multipliy` for the intrinsic is that it describes an operation which is more general than the `fixed_point_multiply` we do in Relay. So `fixed_point_multiply(x,m,s) = qmuls(x, m, 31, s)`. We might call it `q_multiply_shift`? Or `q_multiply_scale`?

anijain2305 replied:
`q_multiply_shift` is better imo compared to `qmuls`

@kparzysz-quic Any suggestions here?

tmoreau89 replied:
Same here, I like `q_multiply_shift`

anijain2305 commented on src/relay/op/tensor/unary.cc:
```diff
@@ -274,6 +274,20 @@ TVM_REGISTER_GLOBAL("relay.op._make.clip").set_body_typed([](Expr a, double a_mi
+    .set_attr<TOpIsStateful>("TOpIsStateful", false)
+    .set_attr<FInferCorrectLayout>("FInferCorrectLayout", ElemwiseArbitraryLayout)
+    .set_attrs_type<FixedPointMultiplyAttrs>()
+    .set_support_level(3);
```
I think level 10 is better here

@tqchen any suggestions here?

giuseros replied:
The reason why it is three is that we could implement requantization through only clipping and/or cast, which are level 3 as well.
Moreover, from here: https://tvm.apache.org/docs/langref/relay_op.html#overview-of-operators, 3 is for additional math operators which seems the right category. However, either way it is fine for me.

anijain2305 commented on src/relay/op/tensor/unary.cc:
```diff
@@ -274,6 +274,20 @@ TVM_REGISTER_GLOBAL("relay.op._make.clip").set_body_typed([](Expr a, double a_mi
+TVM_REGISTER_NODE_TYPE(FixedPointMultiplyAttrs);
+
+RELAY_REGISTER_OP("fixed_point_multiply")
+    .describe(R"code( fixed point multiplication )code" TVM_ADD_FILELINE)
```
No need of space between `(` and `fixed`. Similarly at the end.

anijain2305 commented on src/target/intrin_rule.cc:
```diff
@@ -115,6 +115,51 @@ TVM_REGISTER_GLOBAL("tvm.intrin.rule.default.isinf")
+      x = tir::Select(left_shift != zero, x << cast(hp_dtype, left_shift), cast(hp_dtype, x));
+
+      // 3) Perform the multiplication in higher precision.
+      x = x * y;
```
Do we need to cast y to `hp_dtype`?

giuseros replied:
I think `y` gets automatically promoted to `int64`. However, I guess being explicit doesn't hurt !

anijain2305 commented on src/relay/qnn/op/requantize.cc:
```diff
@@ -153,9 +153,19 @@ Expr RequantizeLower(const Expr& input_tensor, const Expr& input_scale,
+      // the fixed_point_muliply intrinsic
+      scaled_int32_t = (is_upward_rounding ? relay::FixedPointMultiply(
+                                                 scaled_int32_t, fixed_point_multiplier, shift)
+                                           : FixedPointMultiply(scaled_int32_t, double_multiplier,
```
We should remove the UPWARD rounding code from the util non-Relay FixedPointMultiply function as that code will not be executed now.

Also, it is confusing now to why we have two same function names :) We should rename the already existing util function.

giuseros replied:
I changed the utility function with `FixedPointMultiplyToNearest` and removed the conditions on the rounding.

anijain2305 commented on python/tvm/tir/op.py:
```diff
@@ -965,6 +965,34 @@ def popcount(x):
+    """Execute a multiplication between two Q-numbers x and y
+    followed by a right shift s. The mathematical expression is:
+
+       out = round(x*y*2^-s)
```
Maybe we should add a line to explain why there is a multiplication factor of 2 (perhaps rounding)

giuseros replied:
I am not sure I follow. Do you mean the multiplication by `2^-s`?

anijain2305 replied:
Oh I completely misread. This is good.

anijain2305 commented on topi/python/topi/arm_cpu/injective.py:
```diff
@@ -62,9 +62,13 @@ def schedule_injective(outs):
     x = outs[0]
+    ins = x.op.input_tensors
+    dtype = ins[0].dtype if len(ins) > 0 else x.dtype
+    max_vlen = 4 if dtype == 'int32' else 8
```
Seems like 4 should be better for float32 as well. If it is, then maybe we should always use 4 instead of 8.

giuseros replied:
I will set at 4, as a good enough value. The point is that in a network this doesn't matter, as it will get fused to the convolution op (which will decide the level of vectorization)

anijain2305 commented on topi/python/topi/arm_cpu/tensor_intrin.py:
```diff
@@ -451,3 +451,55 @@ def _instr(index):
+
+    # Don't use this intrinsic if we don't have a int32x4 vector
+    # and if we are not multiplying q31 numbers
+    if x.dtype != "int32x4" and q == 31:
```
Can you please double check the condition? Should there be or here?

anijain2305 commented on topi/python/topi/math.py:
```diff
@@ -612,6 +612,31 @@ def _compute(*indices):
+
+    Parameters
+    ----------
+    x :          tvm.te.Tensor or Expr
```
Lets be consistent with the doc format. You can take a look at the example just above this.

giuseros commented:
@anijain2305 , @tmoreau89 , @kparzysz-quic I tried to address/reply to your comments. I added a `test_fixed_point_multiply()` in `test_op_level3.py` to test the intrinsic (which receives a lot of additional testing since it will be used in every `qnn::requantize` operation).

giuseros commented:
Quick update on this. 
* I had to remove checks for input data type to be `int32` from the intrinsic (and TOPI). Indeed, this can be called from `relay/quantize/realize.cc` with `int64` and `int8` input datatypes. After all, we convert upfront to `int64` so the input data type shouldn't matter. 
* I also applied some changes to the intrinsic so that now the casts are more explicit.

anijain2305 commented on topi/python/topi/math.py:
```diff
@@ -612,6 +612,31 @@ def _compute(*indices):
+    Parameters
+    ----------
+    x :          tvm.te.Tensor or Expr
+                 Input argument.
```
Indentation - There should be two tabs here. Seems many tabs.

giuseros replied:
I reformatted it to follow the same style of other help texts.

anijain2305 commented:
@tmoreau89 @kparzysz-quic Please review again and approve explicitly.

tmoreau89 commented:
@giuseros thanks for adding the test case, and thank you @anijain2305 for the thorough review. I suggest to follow through with the naming tweak and this PR should be good to go.

giuseros commented:
Hi @anijain2305 , @tmoreau89 , @kparzysz-quic 
I addressed the name change, reformatted the help text in `math.py` and addressed an integration change in the arm intrinsic (`q.val` -> `q.value`). 

Thank you all for your comments and your time!

tmoreau89 commented:
@anijain2305 please let us know if your requested changes have been addressed! thanks!

anijain2305 commented:
Thanks @giuseros @tmoreau89 @kparzysz-quic  This is merged!
